### PR TITLE
All files

### PR DESCRIPTION
**Pull request by yarinkagal**

### CODE REVIEW

**Comment by yarinkagal** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: API_Handlers/API_Handler.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+        public async Task<HttpResponseMessage> sendGetRequest(string? scanId)
+        {
+            using (HttpClient httpClient = new HttpClient())
+            {
+                // Set the authorization header with the token
+                string accessToken = Auth_Handler.Instance.accessToken;
+                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+
+                // Detect usage - DO NOT MODIFY
```
Comment (shklar) [inline API_Handlers/API_Handler.cs]: We can explain here that this is used for telemetry purposes to identify this specific client

File: API_Handlers/API_Handler.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                httpClient.DefaultRequestHeaders.Add("X-POWERBI-ADMIN-CLIENT-NAME", "FabricScanningClient");
+                
+                // Send the GET request
+                HttpResponseMessage response = await httpClient.GetAsync(apiUriBuilder.Uri + $"/{scanId}");
+
+                int retryAfter = await verifySuccess(response);
+
+                if (retryAfter > 0)
+                {
+                    Console.WriteLine($"To many requests for {apiName} API. Retring in {retryAfter / 1000} seconds");
```
Comment (shklar) [inline API_Handlers/API_Handler.cs]: Too
Retrying

File: API_Handlers/ModifiedAPI_Handler.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace Fabric_Metadata_Scanning
+{
+    public class ModifiedAPI_Handler : API_Handler
+    {
+        public string outputFolder; 
+
+        public ModifiedAPI_Handler() : base("modified")
+        {
+            bool excludeInActiveWorkspaces = Configuration_Handler.Instance.getConfig(apiName, "excludeInActiveWorkspaces").Value<bool>();
```
Comment (shklar) [inline API_Handlers/ModifiedAPI_Handler.cs]: Inactive - It's one word

File: API_Handlers/ModifiedAPI_Handler.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            bool excludePersonalWorkspaces = Configuration_Handler.Instance.getConfig(apiName, "excludePersonalWorkspaces").Value<bool>();
+
+            parameters.Add("excludeInActiveWorkspaces", excludeInActiveWorkspaces);
```
Comment (shklar) [inline API_Handlers/ModifiedAPI_Handler.cs]: same

File: API_Handlers/ScanResultAPI_Handler.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            }
+
+            artifactsCounters = new Dictionary<string, int>()
+            {
+                {"workspaces",0 },
+                {"reports",0 },
+                {"dashboards",0 },
+                {"datasets",0 },
+                {"dataflows",0 },
+                {"datamarts",0 }
```
Comment (shklar) [inline API_Handlers/ScanResultAPI_Handler.cs]: Can we also add Fabric artifacts?

File: API_Handlers/WorkspaceInfoAPI_Handler.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                    workspaces = workspacesToScan
+                };
+
+                string requestJsonString = JsonConvert.SerializeObject(requestBody);
+                HttpContent content = new StringContent(requestJsonString, Encoding.UTF8, "application/json");
+
+                string accessToken = Auth_Handler.Instance.accessToken;
+                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+
+                // Detect usage - DO NOT MODIFY
```
Comment (shklar) [inline API_Handlers/WorkspaceInfoAPI_Handler.cs]: Remove?

File: Auth_Handler.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+namespace Fabric_Metadata_Scanning
+{
+    class Auth_Handler
+    {
+        private static Auth_Handler instance = null;
+        public static Auth_Handler Instance
+        {
+            get
+            {
+                if (instance == null)
```
Comment (shklar) [inline Auth_Handler.cs]: This isn't a thread safe singleton pattern. If you need it to be thread safe please revise

File: Configuration_Handler.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        private Configuration_Handler()
+        {
+
+        }
+
+        public static Configuration_Handler Instance
+        {
+            get
+            {
+                if (instance == null)
```
Comment (shklar) [inline Configuration_Handler.cs]: same

File: ScanningException.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿
+namespace Fabric_Metadata_Scanning
+{
+    public class ScanningException : Exception
+    {
+        public string ReadMeLink = "https://github.com/yarinkagal/Fabric_Metadata_Scanning/blob/master/ReadMe.md";
```
Comment (shklar) [inline ScanningException.cs]: Please update to Microsoft repo

File: ScanningException.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+        public string HelpLink = "https://learn.microsoft.com/en-us/rest/api/power-bi/admin/workspace-info-get-modified-workspaces";
+
+        public ScanningException(string apiName, string errorMessage) :
+            base($"{apiName} has an error: {errorMessage}. Please check it's properties.")
```
Comment (shklar) [inline ScanningException.cs]: its

File: outputs/Results/2023-12-05-124156.json
Comment (shklar) [inline outputs/Results/2023-12-05-124156.json:1]: Is checking in this file intentional?

File: images/add_registretion.png
Comment (shklar) [inline images/add_registretion.png:1]: rename the file to add_registration

**Review by shklar — APPROVED** [review]

File: outputs/Results/2023-12-05-124156.json
Comment (yarinkagal) [reply outputs/Results/2023-12-05-124156.json:1]: Yes, it is basically an example of how the result looks like

File: API_Handlers/WorkspaceInfoAPI_Handler.cs (hunk 6 above)
Comment (yarinkagal) [reply API_Handlers/WorkspaceInfoAPI_Handler.cs]: Moved to external method setHeaders(HttpClient client) in the API_Handler class

**Review by shklar — APPROVED** [review]

**Review by msft-idoma — APPROVED** [review]

**Pull request merged**